packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -0,0 +1,624 @@
+    if (node == null || node == selectedNode) {
+      return;
+    }
+
+    // Un-select the currently selected node if there is one.
+    if (selectedNode != null) {
+      selectedNode.selected = false;
+    }
+    // Select the new selected node.
+    node.selected = true;
+    selectedNode = node;
+
+    _stackFrameSelectedController.add(node.stackFrame);
+  }
+
+  num getFlameChartWidth() {
+    return rows[0].nodes[0].rect.right - _flameChartInset;
+  }
+
+  FlameChartNode getNode(Offset offset) {
+    final int rowIndex = getRowIndexForY(offset.dy);
+    if (rowIndex < 0 || rowIndex >= rows.length) {
+      return null;
+    }
+    return getNodeInRow(rowIndex, offset.dx);
+  }
+
+  FlameChartNode getNodeInRow(int rowIndex, double x) {
+    final row = rows[rowIndex];
+    final nodes = row.nodes;
+
+    FlameChartNode binarySearchForNode() {
+      int min = 0;
+      int max = nodes.length;
+      while (min < max) {
+        int mid = min + ((max - min) >> 1);
+        final node = nodes[mid];
+        if (x >= node.rect.left && x <= node.rect.right) {
+          return node;
+        }
+        if (x < node.rect.left) {
+          max = mid;
+        }
+        if (x > node.rect.right) {
+          min = mid + 1;
+        }
+      }
+      return null;
+    }
+
+    return nodes.isEmpty ? null : binarySearchForNode();
+  }
+
+  int getRowIndexForY(double y) {
+    if (y < flameChartTop) {
+      return -1;
+    }
+    return math.max((y - flameChartTop) ~/ rowHeightWithPadding, 0);
+  }
+}
+
+class FlameChartCanvas extends FlameChart {
+  FlameChartCanvas({
+    @required CpuProfileData data,
+    @required flameChartWidth,
+    @required flameChartHeight,
+  }) : super(
+          data: data,
+          flameChartWidth: flameChartWidth,
+          flameChartHeight: flameChartHeight,
+        ) {
+    _viewportCanvas = ViewportCanvas(
+      paintCallback: _paintCallback,
+      onTap: _onTap,
+      classes: 'ui-details-section cpu-flame-chart',
+    )..element.element.style.overflow = 'hidden';
+
+    _viewportCanvas.setContentSize(flameChartWidth, flameChartHeight);
+
+    _dragScroll.enableDragScrolling(element);
+    _dragScroll.onVerticalScroll = () {
+      _viewportCanvas.rebuild(force: true);
+    };
+
+    element.element.onMouseWheel.listen(_handleMouseWheel);
+  }
+
+  ViewportCanvas _viewportCanvas;
+
+  CoreElement get element => _viewportCanvas.element;
+
+  /// Maximum scroll delta allowed for scrollwheel based zooming.
+  ///
+  /// This isn't really needed but is a reasonable for safety in case we
+  /// aren't handling some mouse based scroll wheel behavior well, etc.
+  final num maxScrollWheelDelta = 20;
+
+  /// Maximum zoom level we should allow.
+  ///
+  /// Arbitrary large number to accommodate spacing for some of the shortest
+  /// events when zoomed in to [_maxZoomLevel].
+  final _maxZoomLevel = 150;
+  final _minZoomLevel = 1;
+  num zoomLevel = 1;
+
+  num get _zoomMultiplier => zoomLevel * 0.003;
```

Comment:
I would think the zoom multiplier and similar should be on the base class not the canvas subclass.

Reply:
done.